### DISCUSSION TITLE
Stricter validation on latest version breaks on existing configs

### DISCUSSION THREAD

**Discussion by MindTooth** [renovatebot/renovate#27468] [Request Help]
### How are you running Renovate?

Self-hosted Renovate

### If you're self-hosting Renovate, tell us what version of Renovate you run.

37.203.2

### If you're self-hosting Renovate, select which platform you are using.

GitLab self-hosted

### Was this something which used to work for you, and then stopped?

It used to work, and then stopped

### Describe the problem

https://github.com/renovatebot/renovate/pull/25218 introduces some checks that now breaks.

`$ renovate-config-validator config.js`:
```
 INFO: Validating config.js
 WARN: Found errors in configuration
       "file": "config.js",
       "warnings": [
         {
           "topic": "Configuration Error",
           "message": "Configuration option `hostRules[0].token` should be a string."
         },
         {
           "topic": "Configuration Error",
           "message": "Invalid `onboardingConfig.onboardingConfig.extends` configuration: value must be a string."
         }
       ]
```

<code>"Configuration option `hostRules[0].token` should be a string."</code> comes from using:
```js
  hostRules: [
    {
      matchHost: "gitlab.<the rest>",
      token: process.env.RENOVATE_TOKEN,
    },
  ],
```

And when `$RENOVATE_TOKEN` is not set, it errors out as the string is empty.

### Relevant debug logs

<details><summary>Logs</summary>

```

Copy/paste the relevant log(s) here, between the starting and ending backticks

```

</details>

### Have you created a minimal reproduction repository?

Placeholder value, please select the correct response from the dropdown

**Comment by rarkins** [comment]
Is your `extends` array also referencing a missing environment variable?

And in your case would you prefer that the solution is for you to populate with dummy env variables before validating, or pass a flag like `--allow-undefined` to the validator and possibly miss problems?

**Reply by MindTooth** [reply]
Doing `export RENOVATE_TOKEN=hello` makes the warning go away.  Maybe have some logic if `process.env.<variable>` is used, tell the user that this might be a false positive.

In a sense, I think that a small doc section for the validation tool is in order.  To maybe describe all these gotchas.  IMHO.

**Reply by HonkingGoose** [reply]
> In a sense, I think that a small doc section for the validation tool is in order. To maybe describe all these gotchas. IMHO.

@MindTooth Did you see [Renovate docs, Config Validation](https://docs.renovatebot.com/config-validation/)? Please open a PR to improve the docs. 😉

**Reply by marco-mybitbv** [reply]
I got the same warning with extends property, but when I make extends a string. He fails also.
Image: ghcr.io/renovatebot/renovate:37.203.2
Yesterday all was fine with version: ghcr.io/renovatebot/renovate:37.203.1

**With array**
```  
onboardingConfig: {
    $schema: "https://docs.renovatebot.com/renovate-schema.json",
    extends: [
        "gitlab>some/url"
    ],
 }
 ```
 
 ```
 WARN: Found errors in configuration
       "file": "config.js",
       "warnings": [
         {
           "topic": "Configuration Error",
           "message": "Invalid `onboardingConfig.onboardingConfig.extends` configuration: value must be a string."
         }
       ]
  ```

**With string**
```  
onboardingConfig: {
    $schema: "https://docs.renovatebot.com/renovate-schema.json",
    extends: "gitlab>some/url"
 }
 ```

 ```
 WARN: Config needs migrating
 Migrate extends from string to array
 WARN: Found errors in configuration
       "file": "config.js",
       "warnings": [
         {
           "topic": "Configuration Error",
           "message": "Invalid `onboardingConfig.onboardingConfig.extends` configuration: value must be a string."
         }
       ]
  ```

**Reply by rarkins** [reply]
I think this one is a bug and maybe requires a revert. I copy pasted one of the onboarding configs above into my config.js and got:

```
 WARN: Found errors in configuration
       "file": "config.js",
       "warnings": [
         {
           "topic": "Configuration Error",
           "message": "Invalid `onboardingConfig.onboardingConfig.extends` configuration: value must be a string."
         }
       ]
```

Note the double `onboardingConfig.onboardingConfig`

**Comment by HonkingGoose** [comment]
Do we need to fix the [documentation](https://docs.renovatebot.com/examples/self-hosting/#usage) or the code for the example:

> ## Usage
> 
> The following example uses the Renovate CLI tool, which you can install by running `npm i -g renovate`.
> 
> If running your own Renovate bot then you will need a user account that Renovate will run as.
> We recommend you create and use a dedicated account for the bot, e.g. name it `renovate-bot` if on your own instance.
> Create and save a PAT for this account.
> 
> Create a Renovate config file, for example:
> 
> ```js
> module.exports = {
>   endpoint: 'https://self-hosted.gitlab/api/v4/',
>   token: '**gitlab_token**',
>   platform: 'gitlab',
>   onboardingConfig: {
>     extends: ['config:recommended'],
>   },
>   repositories: ['username/repo', 'orgname/repo'],
> };
> ```
> 
> Here change the `logFile` and `repositories` to something appropriate.
> Also replace `gitlab-token` value with the one created during the previous step.
> 
> If you're running against GitHub Enterprise Server, then change the `gitlab` values in the example to the equivalent GitHub ones.
> 
> You can save this file as anything you want and then use the `RENOVATE_CONFIG_FILE` environment variable to tell Renovate where to find it.
> 
> Most people use `cron` to schedule when Renovate runs, usually on an hourly schedule.
> 
> ```sh title="Example bash script that you can point cron to"
> #!/bin/bash
> 
> export PATH="/home/user/.yarn/bin:/usr/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:$PATH"
> export RENOVATE_CONFIG_FILE="/home/user/renovate-config.js"
> export RENOVATE_TOKEN="**some-token**" # GitHub, GitLab, Azure DevOps
> export GITHUB_COM_TOKEN="**github-token**" # Delete this if using github.com
> 
> # Renovate
> renovate
> ```
> 
> Save the script file, and run the script manually.
> Only add the script to `cron` after you checked it works.
> 
> 
> !!! note
>     The GitHub.com token as an environment variable is needed to fetch changelogs that are usually hosted on github.com.
>     You don't need to add it if you are already running the bot against github.com, but you do need to add it if you're using GitHub Enterprise Server, GitLab, Azure DevOps, or Bitbucket.
